src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs

```diff
@@ -922,100 +316,29 @@ public LiveReplicationPerformanceCollector.ReplicationPerformanceType GetReplica
-                _cts.Cancel();
-                try
-                {
-                    _connectionOptionsDisposable?.Dispose();
-                }
-                catch (Exception)
-                {
-                }
-                try
-                {
-                    _stream.Dispose();
-                }
-                catch (Exception)
-                {
-                }
-                try
-                {
-                    _tcpClient.Dispose();
-                }
-                catch (Exception)
-                {
-                }
-
-                try
-                {
-                    _connectionOptions.Dispose();
-                }
-                catch
-                {
-                    // do nothing
-                }
-
-                _replicationFromAnotherSource.Set();
-
-                if (_incomingWork != PoolOfThreads.LongRunningWork.Current)
-                {
-                    try
-                    {
-                        _incomingWork?.Join(int.MaxValue);
-                    }
-                    catch (ThreadStateException)
-                    {
-                        // expected if the thread hasn't been started yet
-                    }
-                }
-                _incomingWork = null;
-
-                _cts.Dispose();
-
-                _attachmentStreamsTempFile.Dispose();
-                _replicationFromAnotherSource.Dispose();
-            }
-            finally
-            {
-                try
-                {
-                    releaser?.Dispose();
-                }
-                catch (Exception)
-                {
-                    // can't do anything about it...
-                }
-            }
-        }
-
-        protected void OnFailed(Exception exception, IncomingReplicationHandler instance) => Failed?.Invoke(instance, exception);
-
-        protected void OnDocumentsReceived(IncomingReplicationHandler instance) => DocumentsReceived?.Invoke(instance);
+        protected override void InvokeOnDocumentsReceived() => DocumentsReceived?.Invoke(this);
 
         internal class MergedDocumentReplicationCommand : TransactionOperationsMerger.MergedTransactionCommand
         {
             private readonly long _lastEtag;
             private readonly DataForReplicationCommand _replicationInfo;
-            
+            private readonly ReplicationLatestEtagRequest.ReplicationType _replicationType;
+
             public MergedDocumentReplicationCommand(DataForReplicationCommand replicationInfo, long lastEtag)
             {
                 _replicationInfo = replicationInfo;
                 _lastEtag = lastEtag;
             }
 
+            public MergedDocumentReplicationCommand(DataForReplicationCommand replicationInfo, long lastEtag, ReplicationLatestEtagRequest.ReplicationType replicationType) : this(replicationInfo, lastEtag)
+            {
+                _replicationType = replicationType;
```

Comment:
do we need that? can we override ` SaveSourceEtag` if needed?

Reply:
done